hudi-platform-service/hudi-metaserver/hudi-metaserver-server/src/main/java/org/apache/hudi/metaserver/store/RelationalDBBasedStorage.java

```diff
@@ -0,0 +1,240 @@
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.common.util.CollectionUtils.isNullOrEmpty;
+import static org.apache.hudi.common.util.ValidationUtils.checkState;
+
+/**
+ * Metadata store based on relation database.
+ */
+public class RelationalDBBasedStorage implements MetaserverStorage, Serializable {
+
+  private final WrapperDao tableDao = new WrapperDao.TableDao();
+  private final WrapperDao timelineDao = new WrapperDao.TimelineDao();
+
+  @Override
+  public void initStorage() throws MetaserverStorageException {
+    WrapperDao dao = new WrapperDao("DDLMapper");
+    dao.updateBySql("createDBs", null);
+    dao.updateBySql("createTables", null);
+    dao.updateBySql("createTableParams", null);
+    dao.updateBySql("createPartitions", null);
+    dao.updateBySql("createTableTimestamp", null);
+    dao.updateBySql("createInstant", null);
+    dao.updateBySql("createInstantMetadata", null);
+    dao.updateBySql("createFiles", null);
+  }
+
+  @Override
+  public boolean createDatabase(String db) throws MetaserverStorageException {
+    Map<String, Object> params = new HashMap<>();
+    params.put("databaseName", db);
+    return tableDao.insertBySql("insertDB", params) == 1;
+  }
+
+  @Override
+  public Long getDatabaseId(String db) throws MetaserverStorageException {
+    List<Long> ids = tableDao.queryForListBySql("selectDBId", db);
+    validate(ids, "db " + db);
+    return ids.isEmpty() ? null : ids.get(0);
+  }
+
+  @Override
+  public boolean createTable(Long dbId, Table table) throws MetaserverStorageException {
+    Map<String, Object> params = new HashMap<>();
+    params.put("dbId", dbId);
+    TableBean tableBean = new TableBean(table);
+    params.put("tableBean", tableBean);
+    return tableDao.insertBySql("insertTable", params) == 1;
+  }
+
+  @Override
+  public Table getTable(String db, String tb) throws MetaserverStorageException {
+    Map<String, Object> params = new HashMap<>();
+    params.put("databaseName", db);
+    params.put("tableName", tb);
+    List<TableBean> table = tableDao.queryForListBySql("selectTable", params);
+    validate(table, "table " + db + "." + tb);
+    return table.isEmpty() ? null : table.get(0).toTable();
+  }
+
+  @Override
+  public Long getTableId(String db, String tb) throws MetaserverStorageException {
+    Map<String, Object> params = new HashMap<>();
+    params.put("databaseName", db);
+    params.put("tableName", tb);
+    List<Long> ids = tableDao.queryForListBySql("selectTableId", params);
+    validate(ids, "table " + db + "." + tb);
+    return ids.isEmpty() ? null : ids.get(0);
+  }
+
+  @Override
+  public String createNewTimestamp(long tableId) throws MetaserverStorageException {
+    // todo: support SSS
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMddHHmmss");
```

Comment:
should actually reuse the format from the existing constant. And make millisecond first supported.

Reply:
fix